### PR TITLE
[REVIEW] Fix python column view

### PR DESCRIPTION
**Pull request by brandon-b-miller**
Fixes https://github.com/rapidsai/cudf/issues/5372

### CODE REVIEW

**Comment by GPUtester** [comment]
Please update the changelog in order to start CI tests.

View the gpuCI docs [here](https://docs.rapids.ai/gpuci#open-pr-or-update-pr).

**Comment by brandon-b-miller** [comment]
Ready for initial feedback before proceeding.

File: python/cudf/cudf/core/column/string.py [hunk 1]
```diff
@@ -3987,6 +3987,10 @@ def __cuda_array_interface__(self):
     def _mimic_inplace(self, other_col, inplace=False):
         out = super()._mimic_inplace(other_col, inplace=inplace)
         return out
+    
+    def view(self, dtype):
+        return self.children[1].view(dtype)
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/string.py:3992]: If `self` has an offset this will unnecessary copy construct a new column. We should use `self.base_children[1]` here instead, but you'll need to calculate the offset into it.

File: python/cudf/cudf/core/column/column.py [hunk 2]
```diff
@@ -385,25 +385,26 @@ def copy(self, deep=True):
...
+            The dtype to view the data as
+        
         """
-        params = Column._replace_defaults(self)
-        params.update(kwargs)
-        if "mask" in kwargs and "null_count" not in kwargs:
-            del params["null_count"]
-        return newcls(**params)
+        if self.null_count > 0:
+            raise ValueError("Can not produce a view of a column with nulls")
```
Comment (karthikeyann) [inline python/cudf/cudf/core/column/column.py:399]: why is column with nulls not supported?

Comment (brandon-b-miller) [reply python/cudf/cudf/core/column/column.py:399]: I had some trouble reasoning about this in all cases. For instance, what if we have `cudf.Series([1, None], dtype='int32').view('int64')`?  I am not sure what the null mask would be in this situation. Although really it should allow cases where the dtypes are the same size, even with nulls.

Comment (karthikeyann) [reply python/cudf/cudf/core/column/column.py:399]: null mask would be exactly same (considering both of them have same offset). nullmask is 1 bit per element for any datatype in libcudf. 
I am not sure how offset is handled in python column view.

Comment (brandon-b-miller) [reply python/cudf/cudf/core/column/column.py:399]: I am not sure I follow. If we kept the same mask, we'd have a mask with length 2 referring to a series of length 1, right?

Comment (karthikeyann) [reply python/cudf/cudf/core/column/column.py:399]: Correct. It won't work for different length. it could still support columns of same return length! (it will be useful at few places)

Comment (brandon-b-miller) [reply python/cudf/cudf/core/column/column.py:399]: 👍 I'll update this to require that `self.null_count == 0 and self.dtype.itemsize == dtype.itemsize`

File: python/cudf/cudf/core/column/column.py [hunk 3]
```diff
@@ -395,16 +395,23 @@ def view(self, dtype):
             The dtype to view the data as
         
         """
-        if self.null_count > 0:
-            raise ValueError("Can not produce a view of a column with nulls")
         dtype = np.dtype(dtype)
+
+        if self.null_count > 0:
+            if self.dtype.itemsize != dtype.itemsize:
+                raise ValueError("Can not produce a view of a column with nulls")
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/column.py:402]: Can we make this error more explicit that we can't produce a view of a column with nulls of a different size dtype?

File: python/cudf/cudf/core/column/column.py [hunk 4]
```diff
@@ -395,16 +395,23 @@ def view(self, dtype):
...
+
         if dtype.kind in ('o', 'u', 's'):
             raise TypeError('Bytes viewed as str without metadata is ambiguous')
+
         if self.data.size % dtype.itemsize:
             raise TypeError(
                 "Source data size must be an even multiple of view data size"
             )
-        return as_column(self.data, dtype=dtype)
+        return build_column(self.data, dtype=dtype, mask=self.mask)
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/column.py:414]: Should use `self.base_data`, `self.base_mask`, `self.offset`, and `self.size` here instead to avoid potential unnecessary copy construction.

File: python/cudf/cudf/core/column/string.py [hunk 5]
```diff
@@ -3987,10 +3987,21 @@ def __cuda_array_interface__(self):
...
-        return self.children[1].view(dtype)
 
+    def view(self, dtype):
+        dtype = np.dtype(dtype)
+        if (
+            self.base_children[0][self.offset]
+            * self.base_children[1].dtype.itemsize
+        ) % dtype.itemsize:
+            raise TypeError(
+                "Slice data size must divide evenly into desired dtype size"
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/string.py:3998]: It being a slice is an implementation detail, the error message should be more generic that we got `x` bytes which isn't divisible by `y` dtype itemsize.

File: python/cudf/cudf/core/column/string.py [hunk 6]
```diff
@@ -3987,10 +3987,21 @@ def __cuda_array_interface__(self):
...
+            )
+        else:
+            offset = self.base_children[0][self.offset] * dtype.itemsize
+            return column.build_column(
+                self.base_children[1].data, dtype=dtype, offset=offset
+            )
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/string.py:4004]: Need to catch the case where nullcount isn't 0 and error out

File: python/cudf/cudf/core/column/categorical.py [hunk 7]
```diff
@@ -1161,6 +1161,9 @@ def _mimic_inplace(self, other_col, inplace=False):
 
         return out
 
+    def view(self, dtype):
+        raise NotImplementedError('Can not create a view of a categorical column')
```
Comment (cwharris) [inline python/cudf/cudf/core/column/categorical.py:1165]: Is this absolute, or should we reword this to say something like "categorical column views are currently unsupported." ?

File: python/cudf/cudf/core/column/column.py [hunk 8]
```diff
@@ -385,25 +385,33 @@ def copy(self, deep=True):
...
-        if "mask" in kwargs and "null_count" not in kwargs:
-            del params["null_count"]
-        return newcls(**params)
+        dtype = np.dtype(dtype)
+
+        if self.null_count > 0:
+            if self.dtype.itemsize != dtype.itemsize:
+                raise ValueError("Can not produce a view of a column with nulls")
+            else:
+                pass
```
Comment (cwharris) [inline python/cudf/cudf/core/column/column.py:404]:
```suggestion
        if self.null_count > 0 and self.dtype.itemsize != dtype.itemsize:
            raise ValueError("Cannot produce view of column with different dtype size if the column contains nulls")
```

File: python/cudf/cudf/core/column/string.py [hunk 9]
```diff
@@ -3988,6 +3988,21 @@ def _mimic_inplace(self, other_col, inplace=False):
         out = super()._mimic_inplace(other_col, inplace=inplace)
         return out
 
+    def view(self, dtype):
+        dtype = np.dtype(dtype)
+        if (
+            self.base_children[0][self.offset]
+            * self.base_children[1].dtype.itemsize
+        ) % dtype.itemsize:
```
Comment (cwharris) [inline python/cudf/cudf/core/column/string.py:3996]: Maybe use some variables here to provide context for what value this expression resolves to. Not sure what I'm looking at. :man_shrugging:

File: python/cudf/cudf/core/column/string.py [hunk 10]
```diff
@@ -3988,6 +3988,21 @@ def _mimic_inplace(self, other_col, inplace=False):
...
+            raise TypeError(
+                "Slice data size must divide evenly into desired dtype size"
+            )
+        else:
```
Comment (cwharris) [inline python/cudf/cudf/core/column/string.py:4000]: Can reduce nesting by removing the `else` statement, since the previous block raises an error.

File: python/cudf/cudf/core/column/column.py (hunk 4 above)
Comment (brandon-b-miller) [reply python/cudf/cudf/core/column/column.py:414]: In the process of doing this I realized a case that throws a bit of a wrench in things. If I'm understanding correctly, what we want to do avoid copying by interpreting the view as the base column reinterpreted as the new datatype, as well as by scaling the offset and size according to the ratio of the old to new dtypes. But what do we do when the slice we're trying to view divides evenly into the new datatype, but the underlying base column does not? For example what if we had `col=[1,2,3,4,5]` with `int32` dtype, and we wanted to do something like `col[1:].view('int64')`? Now the fact that we're trying to construct the view as the same base column plus a different offset and size won't work even though logically what we're asking for should work.

Comment (cwharris) [reply python/cudf/cudf/core/column/column.py:414]: When you say logically, you mean without respect to the underlying memory representation?

Comment (kkraus14) [reply python/cudf/cudf/core/column/column.py:414]: So given this is only safe if there's no nulls anyway, we can just calculate a new offset based on the change in dtype size.

Comment (brandon-b-miller) [reply python/cudf/cudf/core/column/column.py:414]: But that might not be a valid number, right? I mean that viewing the last four `int32` type elements of a five element array as `int64` should be able to give me back a valid length two array of `int64`s. But if we try and do this by first converting the base five element array into `int64` and then just scaling the offset it won't work, the offset would be 0.5.

Comment (brandon-b-miller) [reply python/cudf/cudf/core/column/column.py:414]: I think the only way to do this without a copy in any scenario is to do the actual pointer math to find out where the new data needs to actually start and construct a buffer around that integer.

Comment (kkraus14) [reply python/cudf/cudf/core/column/column.py:414]: I think we need to split this into two different paths:
1) dtype size stays the same: We can take the `base_*` approach to avoid copy constructing a new mask unnecessarily.
2) dtype size changes: since we already guaranteed there's no nulls, we can just use pointer arithmetic to point in the data buffer properly

Comment (brandon-b-miller) [reply python/cudf/cudf/core/column/column.py:414]: That sounds like the way forward 👍 .

File: python/cudf/cudf/core/column/column.py (hunk 3 above)
Comment (rgsl888prabhu) [reply python/cudf/cudf/core/column/column.py:402]: And should we also handle things where element is `NAN` and still valid ?

File: python/cudf/cudf/core/column/column.py (hunk 4 above)
Comment (rgsl888prabhu) [reply python/cudf/cudf/core/column/column.py:414]: As null columns are not supported in case of different dtype sizes, we can do this. But for same dtype sizes, we need to use the same offset.

Comment (brandon-b-miller) [reply python/cudf/cudf/core/column/column.py:414]: Are there any gotchas with this approach with respect to overlapping buffers, like w.r.t memory ownership, etc? Suppose I have an array `ary` and a view `view=ary[n:m].view(to_dtype)`, under which there is a different buffer (by python `id`). What happens when `ary` goes out of scope, is `view` still valid?

File: python/cudf/cudf/core/column/column.py (hunk 3 above)
Comment (brandon-b-miller) [reply python/cudf/cudf/core/column/column.py:402]: Is there anything wrong with just letting it get converted to whatever the binary representation happens to translate to? That appears to be what happens with NumPy.

Comment (kkraus14) [reply python/cudf/cudf/core/column/column.py:402]: Nope, that's just the nature of binary representations :)

File: python/cudf/cudf/core/column/column.py (hunk 4 above)
Comment (kkraus14) [reply python/cudf/cudf/core/column/column.py:414]: If you set the `owner` to the `Buffer` we handle reaching inside that `Buffer` and setting the owner appropriately in the Buffer `__init__`

File: python/cudf/cudf/core/column/column.py [hunk 11]
```diff
@@ -397,21 +397,37 @@ def view(self, dtype):
...
+                )
 
-        if self.data.size % dtype.itemsize:
-            raise TypeError(
-                "Source data size must be an even multiple of view data size"
-            )
-        return build_column(self.data, dtype=dtype, mask=self.mask)
+            new_size = (self.base_size - self.offset) * self.dtype.itemsize
+            #import pdb
+            #pdb.set_trace()
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/column.py:428]: Debug code?
```suggestion
```

File: python/cudf/cudf/core/column/string.py [hunk 12]
```diff
@@ -3989,19 +3989,23 @@ def _mimic_inplace(self, other_col, inplace=False):
...
-        else:
-            offset = self.base_children[0][self.offset] * dtype.itemsize
-            return column.build_column(
-                self.base_children[1].data, dtype=dtype, offset=offset
-            )
+        str_byte_offset = self.base_children[0][self.offset]
+        char_dtype_size = self.base_children[1].dtype.itemsize
+        n_bytes_to_view = (
+            self.base_children[1].size - str_byte_offset
+        ) * char_dtype_size
```
Comment (kkraus14) [inline python/cudf/cudf/core/column/string.py:3999]: I think we need `self.base_children[0][self.offset + self.size]` to determine the end bytes of the the characters as opposed to `self.base_children[1].size`. I'm pretty sure the latter is just the base size.

Comment (kkraus14) [reply python/cudf/cudf/core/column/string.py:3999]: Should also add a test of viewing a string column with both a start and end slice.

Comment (brandon-b-miller) [reply python/cudf/cudf/core/column/string.py:3999]: Oh whoops you're right I think this might not include all of the last element if it's not just a single byte. Looking into it.

File: python/cudf/cudf/tests/test_column.py [hunk 13]
```diff
@@ -136,3 +136,155 @@ def test_column_chunked_array_creation():
...
+
+    expect = cudf.Series(cudf.Series(data)._column[3:].view('int8'))
+    got = cudf.Series(str_host_view(data[3:], 'int8'))
+    
+    assert_eq(expect, got)
+
+    expect = cudf.Series(cudf.Series(data)._column[2:5].view('int8'))
+    got = cudf.Series(str_host_view(data[2:5], 'int8'))
+
+    assert_eq(expect, got)
```
Comment (karthikeyann) [inline python/cudf/cudf/tests/test_column.py:290]: new line at end of file.
Here is vscode settings to do it automatically,
```
"files.insertFinalNewline": true,
"files.trimFinalNewlines": true,
```

File: python/cudf/cudf/core/column/column.py [hunk 14]
```diff
@@ -385,25 +385,47 @@ def copy(self, deep=True):
                 children=self.base_children,
             )
 
-    def view(self, newcls, **kwargs):
-        """View the underlying column data differently using a subclass of
-        ColumnBase
+    def view(self, dtype):
+        """
+        View the data underlying a column as different dtype.
```
Comment (karthikeyann) [inline python/cudf/cudf/core/column/column.py:390]: Add documentation of the requirements and limitations of this method.

File: python/cudf/cudf/core/column/string.py [hunk 15]
```diff
@@ -3988,6 +3988,25 @@ def _mimic_inplace(self, other_col, inplace=False):
         out = super()._mimic_inplace(other_col, inplace=inplace)
         return out
 
+    def view(self, dtype):
+        if self.null_count > 0:
+            raise ValueError("Can not produce a view of a column with nulls")
```
Comment (karthikeyann) [inline python/cudf/cudf/core/column/string.py:3993]: This error should specify string column with nulls.

File: python/cudf/cudf/core/column/string.py (hunk 12 above)
Comment (brandon-b-miller) [reply python/cudf/cudf/core/column/string.py]: Updated. I think I have the string test in there, should be one for the start, one for the end and one for the middle.

File: python/cudf/cudf/tests/test_column.py (hunk 13 above)
Comment (brandon-b-miller) [reply python/cudf/cudf/tests/test_column.py]: Thanks! I didn't know you could do this automatically :)

File: python/cudf/cudf/core/column/column.py (hunk 14 above)
Comment (brandon-b-miller) [reply python/cudf/cudf/core/column/column.py:390]: Done

File: python/cudf/cudf/core/column/string.py (hunk 15 above)
Comment (brandon-b-miller) [reply python/cudf/cudf/core/column/string.py]: Done 👍

File: python/cudf/cudf/tests/test_column.py [hunk 16]
```diff
@@ -255,36 +255,29 @@ def test_column_view_nulls_widths_even():
...
+    expect = cudf.Series(data[slc].view("int64"))
+    got = cudf.Series(sr._column[slc].view("int64"))
 
     assert_eq(expect, got)
 
-    expect = cudf.Series(data[1:3])
-    got = cudf.Series(sr._column[1:3].view("int64"))
-
-def test_column_view_string_slice():
+@pytest.mark.parametrize('slc', [slice(3, None), slice(None, 4), slice(2 , 5)])
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_column.py]: Can we add `slice(1,3)` here to test where we slice off from both the beginning and end? `slice(2, 5)` doesn't slice off the end currently.

**Comment by kkraus14** [comment]
@cwharris @karthikeyann can you give this another look when you get the chance?

File: python/cudf/cudf/tests/test_column.py [hunk 17]
```diff
@@ -144,6 +144,156 @@ def test_column_chunked_array_creation():
...
+    ],
+)
+def test_column_view_valid_string_to_numeric(data, to_dtype):
+    def str_host_view(list_of_str):
+        return np.concatenate(
+            [
+                np.frombuffer(s.encode("utf-8"), dtype=to_dtype)
+                for s in list_of_str
+            ]
+        )
```
Comment (kkraus14) [inline python/cudf/cudf/tests/test_column.py:230]: Move this function out of the test and reuse it in the other test below

**Comment by kkraus14** [comment]
@cwharris can you take another pass? I think all of your comments were addressed
